tests/integration/devfile/cmd_devfile_push_test.go

```diff
@@ -32,254 +32,198 @@ var _ = Describe("odo devfile push command tests", func() {
+			})
+		})
 
-				helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), commonVar.Context)
+		When("odo push is executed with json output", func() {
+
+			BeforeEach(func() {
 				helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
+				output = helper.Cmd("odo", "push", "-o", "json", "--project", commonVar.Project).ShouldPass().Out()
+			})
 
-				output := helper.Cmd("odo", "push", "-o", "json", "--project", commonVar.Project).ShouldPass().Out()
-				utils.AnalyzePushConsoleOutput(output)
+			It("should display push output in JSON format", func() {
 
-				// update devfile and push again
-				helper.ReplaceString("devfile.yaml", "name: FOO", "name: BAR")
-				output = helper.Cmd("odo", "push", "-o", "json", "--project", commonVar.Project).ShouldPass().Out()
 				utils.AnalyzePushConsoleOutput(output)
-
 			})
-		})
+			When("update devfile and push again", func() {
 
-		When("pushing devfile without an .odo folder", func() {
+				BeforeEach(func() {
+					helper.ReplaceString("devfile.yaml", "name: FOO", "name: BAR")
+					output = helper.Cmd("odo", "push", "-o", "json", "--project", commonVar.Project).ShouldPass().Out()
+				})
 
-			It("should be able to push based on name passed", func() {
-				helper.CopyExample(filepath.Join("source", "devfiles", "springboot", "project"), commonVar.Context)
-				helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "springboot", "devfile.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
-				output := helper.Cmd("odo", "push", "--project", commonVar.Project, "springboot").ShouldPass().Out()
-				Expect(output).To(ContainSubstring("Executing devfile commands for component springboot"))
-			})
+				It("should display push updated output in JSON format", func() {
 
-			It("should error out on devfile flag", func() {
-				helper.Cmd("odo", "push", "--project", commonVar.Project, "--devfile", "invalid.yaml").ShouldFail()
+					utils.AnalyzePushConsoleOutput(output)
+				})
 			})
-
 		})
 
-		When("not in context directory", func() {
-			It("should push correct component based on --context flag", func() {
-				newContext := helper.CreateNewContext()
-				defer helper.DeleteDir(newContext)
+		When("running odo push outside the context directory", func() {
+			newContext := ""
+			BeforeEach(func() {
+				newContext = helper.CreateNewContext()
+				helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
+
 				helper.Chdir(newContext)
+				output = helper.Cmd("odo", "push", "--context", commonVar.Context).ShouldPass().Out()
+			})
 
-				helper.Cmd("odo", "create", "nodejs", "--project", commonVar.Project, cmpName).ShouldPass()
+			AfterEach(func() {
+				defer helper.DeleteDir(newContext)
+				helper.Chdir(commonVar.Context)
```

Comment:
Why are we changing the `context` when there is neither a nested `when` nor multiple `It` spec.

Reply:
We are changing the context for the test as the discription of the test says odo push --context should work when we are not in the context dir.
AfterEach and BeforeEach contains all the steps that are required for the test and IT spec will only check the condition of passing or failing depending on the expect.